### PR TITLE
Move initializer plugin to the generic apiserver

### PR DESCRIPTION
**Pull request by caesarxuchao**
* Moves `k8s.io/kuberentes/plugin/pkg/admission/initialization` to `k8s.io/apiserver/pkg/admission/plugin/initialization`
* Moves `k8s.io/kubernetes/pkg/kubeapiserver/admission/configuration` to `k8s.io/apiserver/pkg/admission/configuration`
* The initializer plugin used to depend on `k8s.io/kubernetes` because it does a type assertion of `api.Pod`. It tries to skip mirror pod. I converted that code to use the generic accessor pattern.

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@deads2k @p0lyn0mial PTAL. Thank you.

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 1]
```diff
@@ -53,7 +54,7 @@ func NewAdmissionOptions() *AdmissionOptions {
 	options := &AdmissionOptions{
 		Plugins:                &admission.Plugins{},
 		PluginNames:            []string{},
-		RecommendedPluginOrder: []string{lifecycle.PluginName},
+		RecommendedPluginOrder: []string{initialization.PluginName, lifecycle.PluginName},
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Initializers need to be `DefaultOff`.

**Comment by deads2k** [comment]
needs a squash.

I think the dependency tree is ok.  I'm out tomorrow.  @sttts can you watch this one?

/assign sttts

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: I see. I'll remove it.

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: > I see. I'll remove it.

It is still in the recommended order.  There is a separate field `DefaultOff` which turns it off unless it is specifically requested.

[the remark this answers is not in the document]
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Do you mean the lifecycle plugin should run before the initialization plugin, or do you mean this slice is reversely ordered (I don't see why it's reversely ordered)?

[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: > Do you mean the lifecycle plugin should run before the initialization plugin, or do you mean this slice is reversely ordered (I don't see why it's reversely ordered)?

I think the lifecycle plugin should run before the initialization plugin, though I suppose if we agree to split mutators from validators they end up in different chains altogether.

[the remark this answers is not in the document]
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Done.

Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Thanks. Done.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go [hunk 2]
```diff
@@ -183,8 +188,13 @@ func (i *initializer) Admit(a admission.Attributes) (err error) {
 			// Mirror pods are exempt from initialization because they are created and initialized
 			// on the Kubelet before they appear in the API.
 			// TODO: once this moves to REST storage layer, this becomes a pod specific concern
-			if pod, ok := a.GetObject().(*api.Pod); ok && pod != nil {
-				if _, isMirror := pod.Annotations[api.MirrorPodAnnotationKey]; isMirror {
+			if a.GetKind().GroupKind() == v1.SchemeGroupVersion.WithKind("Pod").GroupKind() {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go]: TODO: move this logic out of k8s.io/apiserver

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 3]
```diff
@@ -53,7 +54,8 @@ func NewAdmissionOptions() *AdmissionOptions {
 	options := &AdmissionOptions{
 		Plugins:                &admission.Plugins{},
 		PluginNames:            []string{},
-		RecommendedPluginOrder: []string{lifecycle.PluginName},
+		RecommendedPluginOrder: []string{lifecycle.PluginName, initialization.PluginName},
+		DefaultOffPlugins:      []string{initialization.PluginName},
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: should we really switch this off?

**Comment by sttts** [comment]
Am a bit surprised that intializers are off by default. As they did not exist at all before this PR, it's fine to discuss and maybe change that in a follow-up.

Please squash. Then it's fine.

**Comment by sttts** [comment]
I think this deserves a release-note (at least if we turn initializers on by default for every apiserver).

**Comment by p0lyn0mial** [comment]
> Am a bit surprised that intializers are off by default. As they did not exist at all before this PR

If I understood David correctly. People tend to add more and more admission plugins so maintaining the correct order of plugins becomes a challenge. The idea is to specify a full order and give a possibility for turning the plugins on and off to the users.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go (hunk 2 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go]: @p0lyn0mial do you have thoughts on how to do this? Mabye moving this logic to a separate kube admission plugin?

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 3 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: It's an alpha feature, so should be default off.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: don't we have feature gates for turning off alpha features by default?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/initialization/initialization.go]: nope, I haven't thought much about `initiailzers`

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 3 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: The story of initializers is a bit twisted. If the admission plugin is enabled, the feature gate will be enabled: https://github.com/kubernetes/kubernetes/pull/51436#issuecomment-325250675

**Comment by caesarxuchao** [comment]
Squashed to 2 commits. PTAL. Thank you.

**Comment by deads2k** [comment]
remaining packages are the "tail". approving

**Pull request merged**